================================================================================
# Nama : Achmed Alridho Zulkarnaen
# Batch : RMT-037
# Live Code : 2
================================================================================

In [55]:
# Melakukan koneksi data dari BiqQuery
from google.colab import auth
from google.cloud import bigquery
auth.authenticate_user()
print('Authenticated')

project_id_akun = "alridho-hacktive8"
client = bigquery.Client(project=project_id_akun)

Authenticated


In [56]:
# Menunjukan seluruh tabel beverage_sales untuk di lihat isinya seperti apa
df = client.query('''

SELECT *
FROM `ftds-hacktiv8-project.phase0_ftds_037_rmt.beverage_sales`

''').to_dataframe()

df

invoice_id        date  store_id       store_name   store_city  \
0     INV-58839200013  2023-05-26      6016       Metro Mall     CLARINDA   
1     INV-56953900014  2023-03-24      6016       Metro Mall     CLARINDA   
2     INV-56345300016  2023-03-03      6016       Metro Mall     CLARINDA   
3     INV-32274100025  2020-11-27      6016       Metro Mall     CLARINDA   
4     INV-32274100030  2020-11-27      6016       Metro Mall     CLARINDA   
...               ...         ...       ...              ...          ...   
6009  INV-46502900008  2022-04-14      4119  Peak Provisions  CEDAR FALLS   
6010  INV-35080700016  2021-03-17      4119  Peak Provisions  CEDAR FALLS   
6011  INV-20939700005  2019-07-30      4119  Peak Provisions  CEDAR FALLS   
6012  INV-50461000007  2022-08-18      4119  Peak Provisions  CEDAR FALLS   
6013  INV-51677600017  2022-09-29      4119  Peak Provisions  CEDAR FALLS   

              product_name   manufacturing_company  pack  unit_volume_ml  \
0      OasisGlow Refresher      NibbleNation Group    12              50   
1          PureElixir Zest        MunchWave Snacks    12              50   
2     Sparkling Bliss Brew      BiteBlend Delights    12             750   
3         VelvetChai Charm        OliverCraft Inc.    12             750   
4     FruitFusion Quencher         CrunchCraft Co.    24             375   
...                    ...                     ...   ...             ...   
6009     CocoCrave Delight  TasteTwist Enterprises    12             750   
6010     CocoCrave Delight  TasteTwist Enterprises    12             750   
6011  FruitFusion Quencher         CrunchCraft Co.     6            1750   
6012    FrostedCocoa Savor        OliverCraft Inc.    12             750   
6013    FrostedCocoa Savor        OliverCraft Inc.    12             750   

      unit_sold  unit_retail_price  total_price  volume_sold_liters  
0             4               1.25         5.00                0.20  
1             5               1.30         6.50                0.25  
2             5               1.55         7.75                3.75  
3             6               1.75        10.50                4.50  
4            24               1.22        12.00                9.00  
...         ...                ...          ...                 ...  
6009         12               6.07        72.84                9.00  
6010         24               6.07       145.68               18.00  
6011         18               1.94        34.92               31.50  
6012         12               3.91        46.92                9.00  
6013         12               3.91        46.92                9.00  

[6014 rows x 13 columns]

In [57]:
# Mendapatkan info dari tabel beverage_sales, karena saya ingin mengetahui data typenya
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6014 entries, 0 to 6013
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   invoice_id             6014 non-null   object 
 1   date                   6014 non-null   dbdate 
 2   store_id               6014 non-null   Int64  
 3   store_name             6014 non-null   object 
 4   store_city             6014 non-null   object 
 5   product_name           6014 non-null   object 
 6   manufacturing_company  6014 non-null   object 
 7   pack                   6014 non-null   Int64  
 8   unit_volume_ml         6014 non-null   Int64  
 9   unit_sold              6014 non-null   Int64  
 10  unit_retail_price      6014 non-null   float64
 11  total_price            6014 non-null   float64
 12  volume_sold_liters     6014 non-null   float64
dtypes: Int64(4), dbdate(1), float64(3), object(5)
memory usage: 634.4+ KB


# SQL

## Problem 1 - SQL

1. Kota apa saja yang masuk ke dalam cakupan wilayah distribusi?

In [58]:
df = client.query('''

SELECT DISTINCT store_city AS Kota_Distribusi
FROM `ftds-hacktiv8-project.phase0_ftds_037_rmt.beverage_sales`

''').to_dataframe()

df

Kota_Distribusi
0         CLARINDA
1        MUSCATINE
2       MASON CITY
3  WEST DES MOINES
4          CLINTON
5      CEDAR FALLS

Kota yang masuk ke dalam cakupan wilayah distribusi adalah `Clarinda, Muscatine, Mason City, West Des Moines, Clinton, dan Cedar Falls`. Total terdapat `6 Kota`

2. Berapa rentang harga minimum dan maksimum dari unit minumam yang dijual?

In [59]:
df = client.query('''

SELECT MIN(unit_retail_price) AS minimum_price, MAX(unit_retail_price) AS maximum_price
FROM `ftds-hacktiv8-project.phase0_ftds_037_rmt.beverage_sales`

''').to_dataframe()

df

minimum_price  maximum_price
0           0.98           54.0

Rentang harga minuman yang di jual adalah dari `0.98` hingga `54.0`

3. Bandingkan total penjualan dan jumlah transaksi dari masing-masing kota pada tahun 2022?

In [66]:
df = client.query('''

SELECT store_city AS city, SUM(total_price) AS total_penjualan_2022, SUM(unit_sold) AS jumlah_transaksi_2022, COUNT(invoice_id) AS total_transaksi,
FROM `ftds-hacktiv8-project.phase0_ftds_037_rmt.beverage_sales`
WHERE EXTRACT(YEAR FROM date) IN (2022)
GROUP BY store_city;

''').to_dataframe()

df

city  total_penjualan_2022  jumlah_transaksi_2022  \
0         CLARINDA               7166.26                   4352   
1        MUSCATINE               6983.93                   3384   
2       MASON CITY               1623.60                    564   
3  WEST DES MOINES              37953.48                  10608   
4          CLINTON               9175.91                   6269   
5      CEDAR FALLS              15790.14                   4592   

   total_transaksi  
0              248  
1              234  
2               47  
3               47  
4              387  
5              417

Berikut adalah perbandingan total penjualan dan jumlah transaksi dari masing-masing kota pada tahun 2022

4. Produk apa yang paling banyak diminati pada tahun 2022?

In [34]:
df = client.query('''

SELECT product_name, SUM(unit_sold) AS jumlah_unit_terjual_2022
FROM `ftds-hacktiv8-project.phase0_ftds_037_rmt.beverage_sales`
WHERE EXTRACT(YEAR FROM date) IN (2022)
GROUP BY product_name
ORDER BY jumlah_unit_terjual_2022 DESC
LIMIT 1;
''').to_dataframe()

df

product_name  jumlah_unit_terjual_2022
0  CitrusBurst Breeze                      7415

Produk yang paling banyak diminati pada tahun 2022 adalah `CitrusBurst Breeze` dengan unit terjual sebanyak `7415`

# PANDAS

## Problem 2 PANDAS

In [35]:
import pandas as pd
df = client.query('''SELECT * FROM `ftds-hacktiv8-project.phase0_ftds_037_rmt.beverage_sales`''').to_dataframe()

df

invoice_id        date  store_id       store_name   store_city  \
0     INV-58839200013  2023-05-26      6016       Metro Mall     CLARINDA   
1     INV-56953900014  2023-03-24      6016       Metro Mall     CLARINDA   
2     INV-56345300016  2023-03-03      6016       Metro Mall     CLARINDA   
3     INV-32274100025  2020-11-27      6016       Metro Mall     CLARINDA   
4     INV-32274100030  2020-11-27      6016       Metro Mall     CLARINDA   
...               ...         ...       ...              ...          ...   
6009  INV-46502900008  2022-04-14      4119  Peak Provisions  CEDAR FALLS   
6010  INV-35080700016  2021-03-17      4119  Peak Provisions  CEDAR FALLS   
6011  INV-20939700005  2019-07-30      4119  Peak Provisions  CEDAR FALLS   
6012  INV-50461000007  2022-08-18      4119  Peak Provisions  CEDAR FALLS   
6013  INV-51677600017  2022-09-29      4119  Peak Provisions  CEDAR FALLS   

              product_name   manufacturing_company  pack  unit_volume_ml  \
0      OasisGlow Refresher      NibbleNation Group    12              50   
1          PureElixir Zest        MunchWave Snacks    12              50   
2     Sparkling Bliss Brew      BiteBlend Delights    12             750   
3         VelvetChai Charm        OliverCraft Inc.    12             750   
4     FruitFusion Quencher         CrunchCraft Co.    24             375   
...                    ...                     ...   ...             ...   
6009     CocoCrave Delight  TasteTwist Enterprises    12             750   
6010     CocoCrave Delight  TasteTwist Enterprises    12             750   
6011  FruitFusion Quencher         CrunchCraft Co.     6            1750   
6012    FrostedCocoa Savor        OliverCraft Inc.    12             750   
6013    FrostedCocoa Savor        OliverCraft Inc.    12             750   

      unit_sold  unit_retail_price  total_price  volume_sold_liters  
0             4               1.25         5.00                0.20  
1             5               1.30         6.50                0.25  
2             5               1.55         7.75                3.75  
3             6               1.75        10.50                4.50  
4            24               1.22        12.00                9.00  
...         ...                ...          ...                 ...  
6009         12               6.07        72.84                9.00  
6010         24               6.07       145.68               18.00  
6011         18               1.94        34.92               31.50  
6012         12               3.91        46.92                9.00  
6013         12               3.91        46.92                9.00  

[6014 rows x 13 columns]

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6014 entries, 0 to 6013
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   invoice_id             6014 non-null   object 
 1   date                   6014 non-null   dbdate 
 2   store_id               6014 non-null   Int64  
 3   store_name             6014 non-null   object 
 4   store_city             6014 non-null   object 
 5   product_name           6014 non-null   object 
 6   manufacturing_company  6014 non-null   object 
 7   pack                   6014 non-null   Int64  
 8   unit_volume_ml         6014 non-null   Int64  
 9   unit_sold              6014 non-null   Int64  
 10  unit_retail_price      6014 non-null   float64
 11  total_price            6014 non-null   float64
 12  volume_sold_liters     6014 non-null   float64
dtypes: Int64(4), dbdate(1), float64(3), object(5)
memory usage: 634.4+ KB


1. buat kolom baru, menunjukkan representasi musim dengan ketentuan berikut:

In [40]:
df['date'] = pd.to_datetime(df['date'])

season = []
for bulan in df['date'].dt.month:
  if bulan in [3,4,5]:
    season.append('Spring')
  elif bulan in [6,7,8]:
    season.append('Summer')
  elif bulan in [9,10,11]:
    season.append('Autumn')
  else:
    season.append('Winter')

df['Season'] = season
df

invoice_id       date  store_id       store_name   store_city  \
0     INV-58839200013 2023-05-26      6016       Metro Mall     CLARINDA   
1     INV-56953900014 2023-03-24      6016       Metro Mall     CLARINDA   
2     INV-56345300016 2023-03-03      6016       Metro Mall     CLARINDA   
3     INV-32274100025 2020-11-27      6016       Metro Mall     CLARINDA   
4     INV-32274100030 2020-11-27      6016       Metro Mall     CLARINDA   
...               ...        ...       ...              ...          ...   
6009  INV-46502900008 2022-04-14      4119  Peak Provisions  CEDAR FALLS   
6010  INV-35080700016 2021-03-17      4119  Peak Provisions  CEDAR FALLS   
6011  INV-20939700005 2019-07-30      4119  Peak Provisions  CEDAR FALLS   
6012  INV-50461000007 2022-08-18      4119  Peak Provisions  CEDAR FALLS   
6013  INV-51677600017 2022-09-29      4119  Peak Provisions  CEDAR FALLS   

              product_name   manufacturing_company  pack  unit_volume_ml  \
0      OasisGlow Refresher      NibbleNation Group    12              50   
1          PureElixir Zest        MunchWave Snacks    12              50   
2     Sparkling Bliss Brew      BiteBlend Delights    12             750   
3         VelvetChai Charm        OliverCraft Inc.    12             750   
4     FruitFusion Quencher         CrunchCraft Co.    24             375   
...                    ...                     ...   ...             ...   
6009     CocoCrave Delight  TasteTwist Enterprises    12             750   
6010     CocoCrave Delight  TasteTwist Enterprises    12             750   
6011  FruitFusion Quencher         CrunchCraft Co.     6            1750   
6012    FrostedCocoa Savor        OliverCraft Inc.    12             750   
6013    FrostedCocoa Savor        OliverCraft Inc.    12             750   

      unit_sold  unit_retail_price  total_price  volume_sold_liters  bulan  \
0             4               1.25         5.00                0.20      5   
1             5               1.30         6.50                0.25      3   
2             5               1.55         7.75                3.75      3   
3             6               1.75        10.50                4.50     11   
4            24               1.22        12.00                9.00     11   
...         ...                ...          ...                 ...    ...   
6009         12               6.07        72.84                9.00      4   
6010         24               6.07       145.68               18.00      3   
6011         18               1.94        34.92               31.50      7   
6012         12               3.91        46.92                9.00      8   
6013         12               3.91        46.92                9.00      9   

      Season  
0     Spring  
1     Spring  
2     Spring  
3     Autumn  
4     Autumn  
...      ...  
6009  Spring  
6010  Spring  
6011  Summer  
6012  Summer  
6013  Autumn  

[6014 rows x 15 columns]

Berikut merupakan tabel beverage_sales yang telah di tambahkan column `Season` pada bagian paling kanan

2. Berapa total barang yang terjual kepada toko Peak Provisions sepanjang tahun 2022? (Gunakan pandas query untuk menjawab soal no. 2)

In [54]:
# Karena diperintahkan menggunakan pandas query maka cara yang dilakukan adalah seperti ini, tapi saya juga sudah menaruh cara non-querynya di bawah cara query, cara tersebut yang saya biasa gunakan
df['date'] = pd.to_datetime(df['date'])

peak_provisions = df.query('store_name == "Peak Provisions" and date.dt.year == 2022')
# peak_provisions = df[(df['store_name']=='Peak Provisions') & (df['date'].dt.year == 2022)] -----> Jika ingin menggunakan cara non-query

total_barang_terjual = peak_provisions['unit_sold'].sum()

total_barang_terjual

4592

Total barang yang terjual kepada toko Peak Provisions sepanjang tahun 2022 adalah `4592 Barang`

## Kesimpulan Akhir

Saya telah berhasil dalam melakukan explorasi seluruh data yang ada. Berikut merupakan beberapa hasil data yang saya temukan:
1. Kota unique yang terdapat dalam tabel adalah `Clarinda, Muscatine, Mason City, West Des Moines, Clinton, dan Cedar Falls`
2. Rentang harga minuman yang di jual mulai dari `0.98 - 54.0`
3. Produk yang paling banyak diminati pada tahun 2022 adalah `CitrusBurst Breeze` dengan unit terjual sebanyak `7415`
4. Saya berhasil dalam melakukan analisis `Musim` pada tabel

Dengan ini saya menyimpulkan bahwa saya telah menggunakan query dalam SQL dan Pandas secara baik yang diantaranya saya melakukan manipulasi data dan explorasi data.

Dengan informasi yang sudah saya dapatkan ini juga perusahaan-perusahaan yang mendapatkan data ini nantinya bisa menyimpulkan bahwa produk apa saja sih yang akan laku ke depannya dan dapat melakukan analisis lebih lanjut lagi mengenai produk-produk mereka